webapp/src/main/java/com/box/l10n/mojito/service/cache/DatabaseCacheConfiguration.java

```diff
@@ -0,0 +1,93 @@
+package com.box.l10n.mojito.service.cache;
+
+import com.google.common.base.Preconditions;
+import org.springframework.beans.factory.annotation.Lookup;
+import org.springframework.beans.factory.config.BeanDefinition;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Scope;
+
+import java.time.Duration;
+
+/**
+ * This class enables custom configurations for a database-backed cache.
+ * By default, no TTL is configured (an expiry date isn't writen for entries)
+ * and cache evictions for expired entries are attempted once every 10,000 writes.
+ * The serializer and deserializers for the keys and values are also configurable.
+ *
+ * @author garion
+ */
+@Configuration
+@Scope(BeanDefinition.SCOPE_PROTOTYPE)
+public class DatabaseCacheConfiguration {
+    private double writesToExpiredEvictionRatio;
+    private Duration ttl;
+    private boolean evictEntryOnDeserializationFailure;
+
+    private SerializationPair<Object> keySerializationPair;
+    private SerializationPair<Object> valueSerializationPair;
+
+    public DatabaseCacheConfiguration() {
+        this.writesToExpiredEvictionRatio = 1D / 10_000;
+        this.ttl = Duration.ZERO;
+        this.evictEntryOnDeserializationFailure = true;
+        this.keySerializationPair = new DefaultSerializationPair();
+        this.valueSerializationPair = new DefaultSerializationPair();
+    }
+
+    public void setTtl(Duration ttl) {
+        Preconditions.checkNotNull(ttl);
+        this.ttl = ttl;
+    }
+
+    public DatabaseCache getConfiguredDatabaseCache(String cacheName, DatabaseCacheConfiguration databaseCacheConfiguration) {
+        return this.getDatabaseCache(cacheName, databaseCacheConfiguration);
+    }
+
+    /**
+     * Method-based dependency injection is required here to be able to create multiple {@Link DatabaseCache} instances
+     * at runtime with different configurations that get their own dependencies auto-wired.
+     */
+    @Lookup
```

Comment:
Yes, that's the recommended approach by Spring for this scenario: https://docs.spring.io/spring-framework/docs/current/reference/html/core.html#beans-factory-method-injection


As we can have different caches instantiated for different scenarios (MT, locales, pluralForms etc.) - we need different instances for the Caches. 
The class itself has autowired dependencies - so we can not new() it up manually - those objects need to be created by Spring so that all of the dependencies are resolved correctly. The key thing here is that we do not have only one Cache instantiated - we can have multiple ones configured (as with any cache). I initially attempted to create Bean definitions for each with @Qualifier annotations - however that doesn't help in this scenario.

The cache instances also have different configurations associated with them (name, settings - similar to RedisCache and others) - information which is only available at runtime. The cacheConfig objects similarly need to be separate objects - as different caches will likely use different setting configurations.


Now, the only way I could get all of the above to work is by using the @Lookup annotation provided by Spring to create prototype-scoped instances of our component. (The other option I found was to grab a reference to a BeanFactory to essentially create the components through the Spring factory - however, that leaks Spring-specific code in the classes - which I thought was best avoided).


I couldn't figure out a way to do that using standard constructors - how would it work given the above constraints - while maintaining the ability to have different caches in place that get configured differently at runtime (different names, ttl configs etc.)?

Reply:
just for the record as discussed, @Configurable can be similar to inject field using `new`.

Here are some suggestions. 

1) just be very specific about create a cache for the MTService

```properties
l10n.mtservice.cache.inmemory.ttl=1000
l10n.mtservice.cache.inmemory.max-size=5000
l10n.mtservice.cache.database.enabled=true
l10n.mtservice.cache.database.ttl=5000
l10n.mtservice.cache.database.eviction-cron=* * * * 5
```
```java

@Configuration
@ConfigurationProperties(prefix = "l10n.mtservice.cache")
public class MTServiceCacheConfig {

    Database database = new Database();

    InMemory inMemory = new InMemory();

    static class Database {
        boolean enabled = false;
        int ttl = -1;
        String evictionCron = null;

        public boolean isEnabled() {
            return enabled;
        }

        public void setEnabled(boolean enabled) {
            this.enabled = enabled;
        }

        public int getTtl() {
            return ttl;
        }

        public void setTtl(int ttl) {
            this.ttl = ttl;
        }

        public String getEvictionCron() {
            return evictionCron;
        }

        public void setEvictionCron(String evictionCron) {
            this.evictionCron = evictionCron;
        }
    }

    static class InMemory {
        int ttl = -1;

        public int getTtl() {
            return ttl;
        }

        public void setTtl(int ttl) {
            this.ttl = ttl;
        }
    }

    public Database getDatabase() {
        return database;
    }

    public void setDatabase(Database database) {
        this.database = database;
    }

    public InMemory getInMemory() {
        return inMemory;
    }

    public void setInMemory(InMemory inMemory) {
        this.inMemory = inMemory;
    }
}
```
Then you autowrite that config in CacheConfig and create the cache instance just for the mt service

2) enable configuration of many generic database cache that can be used kind of what spring provides.

```properties
l10n.caches.database.guinness.ttl=100
l10n.caches.database.guinness.eviction-cron=* * * * 5
l10n.caches.database.kilkenny.ttl=200
l10n.caches.database.kilkenny.eviction-cron=* * * * 5
```

```java
package com.box.l10n.mojito.service.mtcache;

import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.context.annotation.Configuration;

import java.util.LinkedHashMap;
import java.util.Map;

@Configuration
@ConfigurationProperties(prefix = "l10n.caches")
public class DatabaseCachesConfig {

    Map<String, DatabaseCache> database = new LinkedHashMap<>();

    static class DatabaseCache {
        int ttl = -1;
        String evictionCron;

        public int getTtl() {
            return ttl;
        }

        public void setTtl(int ttl) {
            this.ttl = ttl;
        }

        public String getEvictionCron() {
            return evictionCron;
        }

        public void setEvictionCron(String evictionCron) {
            this.evictionCron = evictionCron;
        }
    }

    public Map<String, DatabaseCache> getDatabase() {
        return database;
    }

    public void setDatabase(Map<String, DatabaseCache> database) {
        this.database = database;
    }
}

```

You can iterate on the list like that, so basically in the CacheConfig class you can just instanciate/register has many as you want
```
databaseCachesConfig.getDatabase().entrySet().stream().forEach(e ->
                System.out.println(e.getKey() + " : " + e.getValue().getTtl() + "," + e.getValue().getEvictionCron()));
```

Reply:
Thank you for the guidance there - that was very useful.
I've implemented the suggested pattern now.